x-pack/plugin/mapper-match-only-text/src/main/java/org/elasticsearch/xpack/matchonlytext/mapper/MatchOnlyTextFieldMapper.java

```diff
@@ -0,0 +1,289 @@
+        public MatchOnlyTextFieldMapper build(ContentPath contentPath) {
+            FieldType fieldType = new FieldType(Defaults.FIELD_TYPE);
+            fieldType.setStored(store.get());
+            MatchOnlyTextFieldType tft = buildFieldType(fieldType, contentPath);
+            MultiFields multiFields = multiFieldsBuilder.build(this, contentPath);
+            return new MatchOnlyTextFieldMapper(name, fieldType, tft, analyzers.getIndexAnalyzer(), multiFields, copyTo.build(), this);
+        }
+    }
+
+    public static final TypeParser PARSER = new TypeParser((n, c) -> new Builder(n, c.indexVersionCreated(), c.getIndexAnalyzers()));
+
+    public static class MatchOnlyTextFieldType extends StringFieldType {
+
+        private final Analyzer indexAnalyzer;
+        private final TextFieldType textFieldType;
+
+        public MatchOnlyTextFieldType(String name, boolean stored, TextSearchInfo tsi, Analyzer indexAnalyzer, Map<String, String> meta) {
+            super(name, true, stored, false, tsi, meta);
+            this.indexAnalyzer = Objects.requireNonNull(indexAnalyzer);
+            this.textFieldType = new TextFieldType(name);
+        }
+
+        public MatchOnlyTextFieldType(String name, boolean stored, Map<String, String> meta) {
+            super(
+                name,
+                true,
+                stored,
+                false,
+                new TextSearchInfo(Defaults.FIELD_TYPE, null, Lucene.STANDARD_ANALYZER, Lucene.STANDARD_ANALYZER),
+                meta
+            );
+            this.indexAnalyzer = Lucene.STANDARD_ANALYZER;
+            this.textFieldType = new TextFieldType(name);
+        }
+
+        public MatchOnlyTextFieldType(String name) {
+            this(
+                name,
+                false,
+                new TextSearchInfo(Defaults.FIELD_TYPE, null, Lucene.STANDARD_ANALYZER, Lucene.STANDARD_ANALYZER),
+                Lucene.STANDARD_ANALYZER,
+                Collections.emptyMap()
+            );
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        @Override
+        public String familyTypeName() {
+            return TextFieldMapper.CONTENT_TYPE;
+        }
+
+        @Override
+        public ValueFetcher valueFetcher(QueryShardContext context, String format) {
+            return SourceValueFetcher.toString(name(), context, format);
+        }
+
+        private Query toQuery(Query query, QueryShardContext queryShardContext) {
+            Function<LeafReaderContext, CheckedIntFunction<List<Object>, IOException>> valueFetcherProvider = context -> {
+                SourceLookup sourceLookup = new SourceLookup();
+                ValueFetcher valueFetcher = valueFetcher(queryShardContext, null);
+                valueFetcher.setNextReader(context);
+                return docID -> {
+                    try {
+                        sourceLookup.setSegmentAndDocument(context, docID);
```

Comment:
Should we use the `SourceLookup` from `QueryShardContext`? I guess the _source unlikely to be shared but we've been working on consolidating all document access through `QueryShardContext#searchLookup` (#66256).